node/pkg/watchers/aztec/watcher.go

```diff
@@ -0,0 +1,657 @@
+func NewWatcher(
+	chainID vaa.ChainID,
+	networkID watchers.NetworkID,
+	rpcURL string,
+	contractAddress string,
+	msgC chan<- *common.MessagePublication,
+	obsvReqC <-chan *gossipv1.ObservationRequest,
+) *Watcher {
+	return &Watcher{
+		chainID:            chainID,
+		networkID:          string(networkID),
+		rpcURL:             rpcURL,
+		contractAddress:    contractAddress,
+		msgC:               msgC,
+		obsvReqC:           obsvReqC,
+		readinessSync:      common.MustConvertChainIdToReadinessSyncing(chainID),
+		lastProcessedBlock: DefaultStartBlock,
+	}
+}
+
+// Run starts the watcher service and handles the main event loop
+func (w *Watcher) Run(ctx context.Context) error {
+	logger := supervisor.Logger(ctx)
+	logger.Info("Starting Aztec watcher",
+		zap.String("rpc", w.rpcURL),
+		zap.String("contract", w.contractAddress))
+
+	// Create an error channel and ticker
+	errC := make(chan error)
+	defer close(errC)
+
+	// Signal that basic initialization is complete
+	readiness.SetReady(w.readinessSync)
+
+	// Signal to the supervisor that this runnable has finished initialization
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	// Start the single block processing goroutine
+	common.RunWithScissors(ctx, errC, "aztec_events", func(ctx context.Context) error {
+		logger.Info("Starting Aztec event processor")
+
+		for {
+			select {
+			case err := <-errC:
+				logger.Error("Worker error detected", zap.Error(err))
+				return fmt.Errorf("worker died: %w", err)
+
+			case <-ctx.Done():
+				logger.Info("Context done, shutting down")
+				return ctx.Err()
+
+			default:
+				// Wait before processing more blocks
+				time.Sleep(LogProcessingInterval)
+
+				// Check for and process new blocks
+				if err := w.fetchAndProcessBlocks(ctx, logger); err != nil {
+					logger.Error("Error processing blocks", zap.Error(err))
+					// Continue instead of returning to maintain service
+				}
+
+				// Signal readiness
+				readiness.SetReady(w.readinessSync)
+			}
+		}
+	})
+
+	// Wait for context cancellation or error
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
+}
+
+// fetchAndProcessBlocks checks for new blocks and processes them if found
+func (w *Watcher) fetchAndProcessBlocks(ctx context.Context, logger *zap.Logger) error {
+	// Get the latest block number
+	latestBlock, err := w.fetchLatestBlockNumber(logger)
+	if err != nil {
+		return fmt.Errorf("error getting latest block: %w", err)
+	}
+
+	// Only process if there are new blocks
+	if w.lastProcessedBlock >= latestBlock {
+		logger.Debug("No new blocks to process",
+			zap.Int("latest", latestBlock),
+			zap.Int("lastProcessed", w.lastProcessedBlock))
+		return nil
+	}
+
+	// Log that we found new blocks to process
+	logger.Info("Processing new blocks",
+		zap.Int("from", w.lastProcessedBlock+1),
+		zap.Int("to", latestBlock))
+
+	// Process blocks in batches
+	return w.processBlockRange(ctx, logger, w.lastProcessedBlock+1, latestBlock)
```

Comment:
should we process `latestBlock+1` to capture the full range? since the `latestBlock` is not included by default.